src/runtime-rs/crates/hypervisor/src/device_manager.rs

```diff
@@ -0,0 +1,377 @@
+        for dev in self.devices.values() {
+            let mm = dev.lock().await.get_major_minor().await;
+            if mm.0 == major && mm.1 == minor {
+                return Some(dev.clone());
+            }
+        }
+        None
+    }
+
+    async fn find_device_by_bdf(&self, bdf: Option<&String>) -> Option<ArcBoxDevice> {
+        for dev in self.devices.values() {
+            if dev.lock().await.get_bdf().await == bdf {
+                return Some(dev.clone());
+            }
+        }
+        None
+    }
+
+    async fn find_device_by_host_path(&self, host_path: &str) -> Option<ArcBoxDevice> {
+        for dev in self.devices.values() {
+            if host_path == dev.lock().await.get_host_path().await {
+                return Some(dev.clone());
+            }
+        }
+        None
+    }
+
+    async fn get_device_by_id(&self, id: &str) -> Option<ArcBoxDevice> {
+        self.devices.get(id).map(Arc::clone)
+    }
+
+    fn new_device_id(&self) -> Result<String> {
+        for _ in 0..5 {
+            let rand_bytes = rand::RandomBytes::new(8);
+            let id = format!("{:x}", rand_bytes);
+            if self.devices.get(&id).is_none() {
+                return Ok(id);
+            }
+        }
+        Err(anyhow!("ID are exhausted"))
+    }
+
+    // get_and_set_sandbox_block_index retrieves sandbox block index and increments it for
+    // subsequent accesses. This index is used to maintain the index at which a
+    // block device is assigned to a container in the sandbox.
+    fn get_and_set_sandbox_block_index(&mut self) -> Result<u64> {
+        let current_index = self.block_index;
+
+        if !self.released_index.is_empty() {
+            match self.released_index.pop() {
+                Some(index) => Ok(index),
+                None => Err(anyhow!("failed to get block index")),
+            }
+        } else {
+            self.block_index += 1;
+            Ok(current_index)
+        }
+    }
+
+    // unsetSandboxBlockIndex deletes the current sandbox block index from BlockIndexMap.
+    // This is used to recover from failure while adding a block device.
+    fn unset_sandbox_block_index(&mut self, index: u64) -> Result<()> {
+        self.released_index.push(index);
+        self.released_index.sort_by(|a, b| b.cmp(a));
+        Ok(())
+    }
+}
+
+pub fn new_device_info(
+    device: &oci::LinuxDevice,
+    bdf: Option<String>,
+    io_limits: Option<IoLimits>,
+) -> Result<GenericConfig> {
+    let allow_device_type: Vec<&str> = vec!["c", "b", "u", "p"];
+
+    info!(sl!(), "device type:{:?}", device.r#type);
+    if !allow_device_type.contains(&device.r#type.as_str()) {
+        return Err(anyhow!("runtime not support device type {}", device.r#type));
+    }
+
+    if device.path.is_empty() {
+        return Err(anyhow!("container path can not be empty"));
+    }
+
+    let mut file_mode: u32 = 0;
+    let mut uid: u32 = 0;
+    let mut gid: u32 = 0;
+    if device.file_mode.is_some() {
```

Comment:
file_mod/uid/gid can use `unwrap_or(0)` to reduce code.

Reply:
fixed